### PR TITLE
Issue #7094: Add getTopFrecentSiteInfos history API

### LINKED ISSUE #7094: Add getTopFrecentSiteInfos history API
In https://github.com/mozilla/application-services/issues/2163, we're adding an `getTopFrecentSiteInfos` history API to retrieve the the top frecent sites for implementing https://github.com/mozilla-mobile/fenix/issues/10342. We want to integrate this API into AC perhaps adding it to PlacesHistoryStorage so it can be accessed in Fenix.

### PR DESCRIPTION
**Pull request by gabrielluong**
---

Fixes #7094

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by gabrielluong** [comment]
bors try

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt [hunk 1]
```diff
@@ -145,6 +154,17 @@ enum class RedirectSource {
 
 data class PageObservation(val title: String?)
 
+/**
+ * Information about a top frecent site.
```
Comment (jonalmeida) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:158]: I'd expand this a tiny bit explain what "frecent" is.

Comment (gabrielluong) [reply components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:158]: Fixed.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt [hunk 2]
```diff
@@ -68,6 +68,15 @@ interface HistoryStorage : Storage {
         excludeTypes: List<VisitType> = listOf()
     ): List<VisitInfo>
 
+    /**
+     * Returns a list of the top frecent site infos limited by the given number of items
+     * sorted by most to least frecent.
+     *
+     * @param numItems the number of top frecent sites to return in the list.
+     * @return a list of the top frecent site infos sorted by most to least frecent.
```
Comment (grigoryk) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:76]: nit: `a list of [TopFrecentSiteInfo], most frecent first.`

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt [hunk 3]
```diff
@@ -68,6 +68,15 @@ interface HistoryStorage : Storage {
         excludeTypes: List<VisitType> = listOf()
     ): List<VisitInfo>
 
+    /**
+     * Returns a list of the top frecent site infos limited by the given number of items
+     * sorted by most to least frecent.
+     *
+     * @param numItems the number of top frecent sites to return in the list.
```
Comment (grigoryk) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:75]: what happens if there are less frecent sites available than the requested number?

File: components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt [hunk 4]
```diff
@@ -139,6 +139,40 @@ class PlacesHistoryStorageTest {
...
+
+        var infos = history.getTopFrecentSites(0)
+        assertEquals(0, infos.size)
+
+        infos = history.getTopFrecentSites(3)
+        assertEquals(3, infos.size)
+        assertEquals("https://www.mozilla.com/foo/bar/baz", infos[0].url)
+        assertEquals("https://www.example.com/123", infos[1].url)
+
+        infos = history.getTopFrecentSites(5)
```
Comment (grigoryk) [inline components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt:167]: can you also add a test case with more than `toAdd.size` items requested? Like, 100.

File: docs/changelog.md [hunk 5]
```diff
@@ -15,6 +15,10 @@ permalink: /changelog/
 * **browser-icons**
   * Fixed issue [#7142](https://github.com/mozilla-mobile/android-components/issues/7142)
 
+* **browser-storage-sync**
+  * Added `getTopFrecentSites` to `PlacesHistoryStorage` to returns a list of the top frecent site infos
```
Comment (grigoryk) [inline docs/changelog.md:19]: spelling: "which returns".

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt (hunk 2 above)
Comment (gabrielluong) [reply components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:76]: Fixed.

File: docs/changelog.md (hunk 5 above)
Comment (gabrielluong) [reply docs/changelog.md:19]: Fixed.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt (hunk 3 above)
Comment (gabrielluong) [reply components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:75]: In the SQL query, we do a limit based on the provided numItems. If there isn't enough entries to meet the requested number, it should still do the standard SQL operation of providing the entries it does have.

File: components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt (hunk 4 above)
Comment (gabrielluong) [reply components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt:167]: Added

**Comment by gabrielluong** [comment]
bors try

**Comment by gabrielluong** [comment]
bors r=jonalmeida,grigoryk

**Pull request merged**